### PR TITLE
Rockchip: initial support

### PR DESCRIPTION
**Pull request by Kwiboo**
This PR adds initial support for Rockchip SoCs (video HW decoding not included).

~~This include commits from #2100 and #2108 and will be rebased once they have been merged into master.~~
~~There is also a temporary commit to fix a aarch64 build issue caused by `pfcd_hevc_optimisations.patch` that will be dropped once a proper solution is merged into master.~~

There are some issues to resolve / work on before this is ready to be merged:
* [x] Swtich to mainline u-boot ~~and drop the project specific package~~
* [x] Drop the use of device specific mkimage scripts
* [x] LED control is no longer working using the latest RK 4.4 kernel
* [x] Update and combine linux patches
* [x] ~~Possible WiFi issue on 64/32-bit build~~

Big thanks to @LongChair, @omegamoon, @Raybuntu and Rockchip engineers for great collaboration working on this.

### CODE REVIEW

**Comment by LongChair** [comment]
Awesome work @Kwiboo ! This has been a long time work :) also special thanks to @yanghanxing and his team for the great support he's been bringing us !

**Comment by Raybuntu** [comment]
Great job @Kwiboo!

File: packages/graphics/mali-rockchip/package.mk [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+PKG_ARCH="arm aarch64"
+PKG_LICENSE="nonfree"
+PKG_SITE="https://github.com/rockchip-linux/libmali"
+PKG_URL="https://github.com/rockchip-linux/libmali/archive/$PKG_VERSION.tar.gz"
+PKG_SOURCE_DIR="libmali-$PKG_VERSION*"
+PKG_SECTION="graphics"
+PKG_SHORTDESC="mali-rockchip: OpenGL ES user-space binary for the ARM Mali GPU family"
+PKG_LONGDESC="mali-rockchip: OpenGL ES user-space binary for the ARM Mali GPU family"
+
+PKG_IS_ADDON="no"
```
Comment (CvH) [inline packages/graphics/mali-rockchip/package.mk]: rm `PKG_IS_ADDON="no"` and change to
```
PKG_SHORTDESC=""
PKG_LONGDESC=""
PKG_AUTORECONF="no"
```

File: packages/lang/gcc-linaro-aarch64-linux-gnu/package.mk [hunk 2]
```diff
@@ -0,0 +1,41 @@
+################################################################################
+#      This file is part of LibreELEC - https://libreelec.tv
+#      Copyright (C) 2016 Team LibreELEC
```
Comment (CvH) [inline packages/lang/gcc-linaro-aarch64-linux-gnu/package.mk]: 2016-present

File: packages/lang/gcc-linaro-aarch64-linux-gnu/package.mk [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+PKG_LICENSE="GPL"
+PKG_SITE=""
+PKG_URL="https://releases.linaro.org/components/toolchain/binaries/7.1-2017.08/aarch64-linux-gnu/gcc-linaro-${PKG_VERSION}-x86_64_aarch64-linux-gnu.tar.xz"
+PKG_SOURCE_DIR="gcc-linaro-${PKG_VERSION}-x86_64_aarch64-linux-gnu"
+PKG_SECTION="lang"
+PKG_SHORTDESC=""
+PKG_LONGDESC=""
+
+PKG_IS_ADDON="no"
+PKG_AUTORECONF="no"
```
Comment (CvH) [inline packages/lang/gcc-linaro-aarch64-linux-gnu/package.mk]: rm `PKG_IS_ADDON="no"`
```
PKG_SHORTDESC=""
PKG_LONGDESC=""
PKG_AUTORECONF="no"
```

File: packages/multimedia/rkmpp/package.mk [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+PKG_SITE="https://github.com/rockchip-linux/mpp"
+PKG_URL="https://github.com/rockchip-linux/mpp/archive/$PKG_VERSION.tar.gz"
+PKG_SOURCE_DIR="mpp-$PKG_VERSION*"
+PKG_DEPENDS_TARGET="toolchain libdrm"
+PKG_SECTION="multimedia"
+PKG_SHORTDESC="rkmpp: Rockchip Media Process Platform (MPP) module"
+PKG_LONGDESC="rkmpp: Rockchip Media Process Platform (MPP) module"
+
+PKG_IS_ADDON="no"
+PKG_AUTORECONF="no"
```
Comment (CvH) [inline packages/multimedia/rkmpp/package.mk]: rm `PKG_IS_ADDON="no"`
```
PKG_SHORTDESC=""
PKG_LONGDESC=""
PKG_AUTORECONF="no"
```

File: packages/multimedia/rkmpp/package.mk [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+if [ "$UBOOT_SOC" = "rk3328" ]; then
+  ENABLE_VP9D="ON"
```
Comment (CvH) [inline packages/multimedia/rkmpp/package.mk]: `ENABLE_VP9D` should be `PKG_ENABLE_VP9D` as we try to rework every custome var to PKG_... (same change `DENABLE_VP9D=$ENABLE_VP9D`)

File: packages/network/rfkill/package.mk [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+PKG_LICENSE="GPL"
+PKG_SITE="https://wireless.wiki.kernel.org/en/users/documentation/rfkill"
+PKG_URL="https://www.kernel.org/pub/software/network/rfkill/$PKG_NAME-$PKG_VERSION.tar.xz"
+PKG_DEPENDS_TARGET="toolchain"
+PKG_SECTION="network"
+PKG_SHORTDESC="rfkill: userspace tool to query the state of the rfkill switches, buttons and subsystem interfaces"
+PKG_LONGDESC="rfkill is a small userspace tool to query the state of the rfkill switches, buttons and subsystem interfaces."
+
+PKG_IS_ADDON="no"
+PKG_AUTORECONF="no"
```
Comment (CvH) [inline packages/network/rfkill/package.mk]: rm `PKG_IS_ADDON="no"`
```
PKG_SHORTDESC=""
PKG_LONGDESC=""
PKG_AUTORECONF="no"
```

File: packages/tools/rkbin/package.mk [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+PKG_LICENSE="nonfree"
+PKG_SITE="https://github.com/rockchip-linux/rkbin"
+PKG_URL="https://github.com/rockchip-linux/rkbin/archive/$PKG_VERSION.tar.gz"
+PKG_SOURCE_DIR="rkbin-$PKG_VERSION*"
+PKG_SECTION="tools"
+PKG_SHORTDESC="rkbin: Rockchip Firmware and Tool Binaries"
+PKG_LONGDESC="rkbin: Rockchip Firmware and Tool Binaries"
+
+PKG_IS_ADDON="no"
+PKG_AUTORECONF="no"
```
Comment (CvH) [inline packages/tools/rkbin/package.mk]: rm `PKG_IS_ADDON="no"`
```
PKG_SHORTDESC=""
PKG_LONGDESC=""
PKG_AUTORECONF="no"
```

File: projects/Rockchip/devices/ROCK64/scripts/mkimage [hunk 8]
```diff
@@ -0,0 +1,48 @@
+################################################################################
+#      This file is part of LibreELEC - https://LibreELEC.tv
```
Comment (CvH) [inline projects/Rockchip/devices/ROCK64/scripts/mkimage]: `https://LibreELEC.tv` -> `https://libreelec.tv`

File: projects/Rockchip/devices/ROCK64/scripts/mkimage [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+#      Copyright (C) 2017 Team LibreELEC
```
Comment (CvH) [inline projects/Rockchip/devices/ROCK64/scripts/mkimage]: 2017-present

File: projects/Rockchip/devices/TinkerBoard/scripts/mkimage [hunk 10]
```diff
@@ -0,0 +1,40 @@
+################################################################################
+#      This file is part of LibreELEC - https://LibreELEC.tv
+#      Copyright (C) 2017 Team LibreELEC
```
Comment (CvH) [inline projects/Rockchip/devices/TinkerBoard/scripts/mkimage]: 2017-present

File: projects/Rockchip/devices/TinkerBoard/scripts/mkimage [hunk 11]
```diff
@@ -0,0 +1,40 @@
+################################################################################
+#      This file is part of LibreELEC - https://LibreELEC.tv
```
Comment (CvH) [inline projects/Rockchip/devices/TinkerBoard/scripts/mkimage]: https://libreelec.tv

File: projects/Rockchip/packages/rtl8723b_bt/package.mk [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="rtl8723b_bt"
+PKG_VERSION=""
```
Comment (CvH) [inline projects/Rockchip/packages/rtl8723b_bt/package.mk]: pls gave it some version or we get `0.0invaild` instead

File: projects/Rockchip/packages/rtl8723b_bt/package.mk [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+PKG_LICENSE="nonfree"
+PKG_SITE="https://github.com/rockchip-linux/rk-rootfs-build/tree/master/overlay-firmware/lib/firmware/rtlbt"
+PKG_URL=""
+PKG_DEPENDS_TARGET="rfkill"
+PKG_SECTION="firmware"
+PKG_SHORTDESC="rtl8723b_bt firmware"
+PKG_LONGDESC="rtl8723b_bt firmware"
+
+PKG_IS_ADDON="no"
+PKG_AUTORECONF="no"
```
Comment (CvH) [inline projects/Rockchip/packages/rtl8723b_bt/package.mk]: rm `PKG_IS_ADDON="no"` and change to
```
PKG_SHORTDESC=""
PKG_LONGDESC=""
PKG_AUTORECONF="no"
```

File: packages/lang/gcc-linaro-aarch64-linux-gnu/package.mk (hunk 3 above)
Comment (Raybuntu) [reply packages/lang/gcc-linaro-aarch64-linux-gnu/package.mk]: I've fixed those in my PR so a rebase should be sufficient :)

**Comment by Kwiboo** [comment]
Rebased on latest master and updated #2100 commits, fixes for @CvH's comments have been squashed.
Also dropped a commit no longer needed after a small update to the devices specific `mkimage` scripts.

**Comment by Kwiboo** [comment]
Updated to latest RK 4.4 kernel, u-boot v2017.11-rc3 and linux patches is now merged files matching my [linux-rockchip topic branches](https://github.com/Kwiboo/linux-rockchip/branches/all).

This should be ready to merge after a final boot test of new clean builds.

**Comment by Kwiboo** [comment]
Clean builds boots and starts kodi, test images:

[LibreELEC-ROCK64.aarch64-9.0-devel-20171102001520-r27249-g4f038a0.img.gz](https://openpht.blob.core.windows.net/kwiboo/LibreELEC-ROCK64.aarch64-9.0-devel-20171102001520-r27249-g4f038a0.img.gz)
[LibreELEC-ROCK64.arm-9.0-devel-20171102034700-r27249-g4f038a0.img.gz](https://openpht.blob.core.windows.net/kwiboo/LibreELEC-ROCK64.arm-9.0-devel-20171102034700-r27249-g4f038a0.img.gz)
[LibreELEC-TinkerBoard.arm-9.0-devel-20171102020010-r27249-g4f038a0.img.gz](https://openpht.blob.core.windows.net/kwiboo/LibreELEC-TinkerBoard.arm-9.0-devel-20171102020010-r27249-g4f038a0.img.gz)

**Comment by Kwiboo** [comment]
Thanks! :+1:

**Comment by chewitt** [comment]
@Kwiboo nice work! .. and big thanks to all the others that helped move this forwards :)

**Pull request merged**